lib/Auth/ApiAuth.php

```diff
@@ -11,11 +11,23 @@
 
 namespace Mautic\Auth;
 
+use Psr\Http\Client\ClientInterface;
+
 /**
  * OAuth Client modified from https://code.google.com/p/simple-php-oauth/.
  */
 class ApiAuth
 {
+    /**
+     * @var \Psr\Http\Client\ClientInterface|null
+     */
+    protected $client;
+
+    public function __construct(ClientInterface $client = null)
```

Comment:
Is there any way this library will work without the Client object? I'm wondering whether making it nullable makes sense.

Reply:
Yes, it will use auto-discovery to find an HTTP client when you pass `null`.

Reply:
Interesting. It feels like magic. I don't like magic in code personally. But the property must be non-nullable, right?

Reply:
If we make it non-nullable, the user would be required to provide an HTTP client and they can't use auto-discovery. Or they would have to set up auto-discovery themselves. I've set up auto-discovery in the auth classes (in AbstractAuth to be precise) as these are the classes that actually need the HTTP client. However, these classes are usually instantiated with this factory class so I also had to add this argument and pass it through. I can move the auto-discovery set up to this class and make the HTTP client non-nullable in the auth classes if that feels less magic to you?

Reply:
What I want to avoid is to users of this library think that it'll work without any PSR HTTP Client. And that's what the nullable argument suggests.

Reply:
The auto-discover in the class makes sense IMO.

@escopecz see also e.g. https://dev.to/rocksheep/how-to-use-psr-http-standards-to-upgrade-your-code-3l92#:~:text=)%3B%0A%20%20%20%20%7D%0A%7D-,Constructor,Discover%20classes%20to%20find%20usable%20classes%20like%20Guzzle%2C%20HTTPPlug%20or%20cURL.,-sendRequest